<a href="https://colab.research.google.com/github/tat5603-prog/Trego_AI_Portfolio/blob/main/Decision_Tree.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier


from sklearn.metrics import (
    accuracy_score,
    confusion_matrix,
    classification_report
)

In [ ]:
# ------------------------------------------------------------
# STEP 3: Load the Kaggle Dataset
# ------------------------------------------------------------

file_path = "/content/drive/My Drive/datasets/student_exam_performance.csv"

df = pd.read_csv(file_path)

print("Dataset loaded successfully!")
print("Number of rows:", len(df))
print("Number of columns:", len(df.columns))

Dataset loaded successfully!
Number of rows: 100000
Number of columns: 44


In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

# Added precision_score and recall_score to the import statement
from sklearn.metrics import (
    accuracy_score,
    confusion_matrix,
    classification_report,
    precision_score,
    recall_score
)


# -----------------------------------
# 1. Load the Kaggle dataset
# ------------------------------------

df = pd.read_csv(file_path)

print(df.head())


# -----------------------------------
# 2. Select Features
# -----------------------------------

X = df[["previous_exam_score", "attendance_percentage", "study_hours_per_day"]]


# -----------------------------------
# 3. Select Target
# -----------------------------------

y = df["pass_status"]


# -----------------------------------
# 4. Handle Missing Values
# -----------------------------------

X["attendance_percentage"] = X["attendance_percentage"].fillna(X['attendance_percentage'].median())


# -----------------------------------
# 5. Split the Data
# -----------------------------------

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)


# -----------------------------------
# 6. Create Decision Tree
# -----------------------------------

model = DecisionTreeClassifier(
    max_depth=3,
    random_state=42
)


# -----------------------------------
# 7. Train the Model
# -----------------------------------

model.fit(X_train, y_train)


# -----------------------------------
# 8. Make Predictions
# -----------------------------------

y_pred = model.predict(X_test)


# -----------------------------------
# 9. Accuracy
# -----------------------------------

accuracy = accuracy_score(y_test, y_pred)

print("\nAccuracy:", accuracy)


# -----------------------------------
# 10. Confusion Matrix
# -----------------------------------

cm = confusion_matrix(y_test, y_pred)

print("\nConfusion Matrix:")
print(cm)


# -----------------------------------
# 11. Precision
# -----------------------------------

precision = precision_score(y_test, y_pred, pos_label='Pass') # Added pos_label to correctly calculate precision for 'Pass' class

print("\nPrecision:", precision)


# -----------------------------------
# 12. Recall
# -----------------------------------

recall = recall_score(y_test, y_pred, pos_label='Pass') # Added pos_label to correctly calculate recall for 'Pass' class

print("\nRecall:", recall)


# -----------------------------------
# 13. Predict a New Passenger
# -----------------------------------

new_passenger_data = {
    'previous_exam_score': [3],
    'attendance_percentage': [25],
    'study_hours_per_day': [10] # Corrected features to match model training input
}

new_passenger_df = pd.DataFrame(new_passenger_data)

prediction = model.predict(new_passenger_df)

if prediction[0] == 'Pass': # Changed from 1 to 'Pass' as per target values
    print("\nPrediction: Passed")
else:
    print("\nPrediction: Failed")

   student_id  age  gender education_level school_type family_income  \
0  STU_000001   20  Female     High School      Public        Middle   
1  STU_000002   17    Male     High School      Public        Middle   
2  STU_000003   18   Other   Undergraduate      Public        Middle   
3  STU_000004   20    Male     High School      Public        Middle   
4  STU_000005   16  Female     High School     Private          High   

  parent_education urban_rural  previous_exam_score  previous_gpa  ...  \
0              NaN       Rural                78.36          2.95  ...   
1              NaN    Suburban                73.40          2.89  ...   
2           Master       Urban                82.76          3.35  ...   
3      High School    Suburban                60.61          2.57  ...   
4         Bachelor    Suburban                74.79          2.84  ...   

   exam_difficulty  exam_preparation_days questions_attempted  \
0           Medium                     27                

/tmp/ipykernel_1763/2128580034.py:44: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X["attendance_percentage"] = X["attendance_percentage"].fillna(X['attendance_percentage'].median())



Accuracy: 0.7949

Confusion Matrix:
[[  668  3796]
 [  306 15230]]

Precision: 0.8004835488279197

Recall: 0.9803038105046344

Prediction: Passed


Part 1:
https://www.kaggle.com/datasets/mobeenfatimah/student-exam-performance-and-success-dataset/data

The data set is called "Student Exam Performance and Success Data Set" The dataset has 44 columns, or features, and 100000 rows. This dataset includes data from 100000 student records and includes information like previous grades, age, and school type. The model will predict whether a student will pass or fail the next exam.

Part 2:

Each row represents a student and some of the important columns are age, previous exam score, previous gpa, attendance percentage, and amount of hours studied. These important columns will be useful in predicting whether the student will pass or fail. The target variable is whether the students pass or fail. 77 percent of students n the dataset passed and 23 percent of the students failed. There are some missing values in some of the columns, like the attendance status. Some columns should not be used for the prediction like the students id number since it does not affect how the student will do on the exam.

Part 3:

The features used in this model are age, previous exam score, previous gpa, attendance percentage, and amount of hours studied. The target is whether the student will pass or fail. I selected these features because they are some of the most impactful on how someone will perform on an exam.

Part 4:

There are missing values in previous gpa and the attendance percentage of the students. In order to solve the problem of missing values we replace the missing value with the average value from the rest of the data. So the values of the missing gpas become the average gpa from the other students. Any unecessary columns will not be used in order to predict how the student will do on the exam.

Part 5:

The testing data needs to be kept seperate from the training data because if the model already knows the exact correct answer for the data its tested on it will just return the pass and fail value it already knows. Testing it on data it has not seen before prevents this and assures the model is not just returning information we gave it. I used 20 percent testing data and 80 percent training. This means 80000 students data was used for training while 20000 students data were used for testing.

Part 6:
A decision tree is appropriate for my model because the model has to look at different aspects and make a decision based on the data provided.

Part 7:

The models accuracy was 0.7949 which is fairly good. The model was able to succesfully identify 668 true positives and 15230 true negatives. The model also predicted 3796 false positives and 306 false negatives. The percision percentage is the number of true positives over the number of true positives added to the number of false positives, meaning the percision of this model is 0.8004. Recall is the number of true positives over the number of true positives plus the number of false negatives. This means the recall for this model was 0.9803. The metric that is most important for this particular problem is the percision as it would be worse if the model told someone they would pass only for them to fail.

Part 8:

One hypothetical new observation could be if the student will get a 90 or higher on the exam using the same features from the data.

Part 9:

My model predicted whether students where going to pass or fail an exam. The model performed fairly well as it scored an 79% in accuracy. I learned from the confusion matrix how many times the model guessed incorrectly and whether it guessed that someone would pass or fail. A first limitation of the model are the missing data that we had add in ourselves. This makes the model less accurate as it does not have the real data to base off of. A second limitation of the model is holding all of the aspects to the same weight. The model is taking into account the persons age just as much as it is taking into account the number of hours the person has studied. This does not make logical sense and would cause the model to be innaccurate. If I had more time I would take into account the fact that the longer a person studied is more impoartant than their age or the grades they have gotten in the past.




